### PR TITLE
[SLDEV-4714] Create PR BSID from jenkins plugin

### PR DESCRIPTION
**Pull request by RoniSegal**

### CODE REVIEW

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIHandler.java [hunk 1]
```diff
@@ -112,7 +112,8 @@ public void setCommandArgument(AbstractCommandArgument commandArgument) {
     private AbstractUpgradeManager getRelevantUpgradeManager(
             UpgradeProxy upgradeProxy, UpgradeConfiguration upgradeConfiguration) {
 
-        boolean useScannerUpgradeManager = CommandModes.Config.equals(commandArgument.getMode());
+        boolean useScannerUpgradeManager =
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIHandler.java]: Not sure that I get it. Mind elaborating?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner.java [hunk 2]
```diff
@@ -191,6 +195,15 @@ private void validateConfigMode() {
         }
     }
 
+    private void validatePrConfigMode(CommandMode.PrConfigView comandMode) {
+        if (StringUtils.isNullOrEmpty(comandMode.getAppName()) || StringUtils.isNullOrEmpty(comandMode.getLatestCommit()) ||
+                StringUtils.isNullOrEmpty(comandMode.getPullRequestNumber()) || StringUtils.isNullOrEmpty(comandMode.getRepoUrl())) {
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner.java]: You are not checking for target branch. Is that intentional?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java [hunk 3]
```diff
@@ -246,7 +246,25 @@ public void setBranches(List<BeginAnalysis> branches) {
         public ConfigView(String appName, String branchName,
                           CommandBuildName buildName, String labId, String buildSessionId, String additionalArguments,
                           List<TechnologyOptions> techOptions) {
-            super(CommandModes.Config, buildSessionId, additionalArguments);
+                          this(appName, branchName, buildName, labId, buildSessionId, additionalArguments,
+                           techOptions, CommandModes.Config);
+        }
+
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Not directly related to this code, just saw in Github: we have security issues with our plugin - one of our dependencies is in risk.
Can you please check if that's easy to upgrade?
https://github.com/jenkinsci/sealights-plugin/network/alert/pom.xml/com.fasterxml.jackson.core:jackson-databind/open

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/CommandModes.java [hunk 4]
```diff
@@ -28,6 +28,11 @@
         @Override public String getDisplayName() {
             return "Config - Create build session data";
         }
+    },
+    PrConfig("PrConfig") {
+        @Override public String getDisplayName() {
+            return "PrConfig - Create build session data for PR integration";
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/CommandModes.java]: Please show this text to Jason and get his approval.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package io.sealights.plugins.sealightsjenkins.buildsteps.cli.entities;
+
+import hudson.util.DescribableList;
+import io.sealights.plugins.sealightsjenkins.buildsteps.cli.configurationtechnologies.TechnologyOptions;
+import io.sealights.plugins.sealightsjenkins.buildsteps.cli.configurationtechnologies.TechnologyOptionsDescriptor;
+
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java]: Please add documentation for all new classes.
Make sure to explain their role.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        this.pullRequestNumber = pullRequestNumber;
+        this.repoUrl = repoUrl;
+        this.targetBranch = targetBranch;
+    }
+
+    public String getLatestCommit() {
+        return latestCommit;
+    }
+
+    public void setLatestCommit(String latestCommit) {
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java]: Can't we use lombok for these classes?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    protected String getCommandName() {
+        return "-prConfig";
+    }
+
+    @Override
+    protected void addBaseArgumentsLine(List<String> commandsList) {
+        if (baseArgs.getTokenData() != null) {
+            addArgumentKeyVal("token", baseArgs.getTokenData().getToken(), commandsList);
+        } else {
+            addArgumentKeyVal("tokenfile", baseArgs.getTokenFile(), commandsList);
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java]: Are you sure you'll always have token file in this case?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        addArgumentKeyVal("targetBranch", prConfigCommandArguments.getTargetBranch(), commandsList);
+
+    }
+
+    @Override
+    public void addAdditionalArguments(List<String> commandsList) {
+        if(prConfigCommandArguments.getTechOptions().size() > 0){
+           addPackagesIncluded(commandsList);
+        }
+        commandsList.add("-enableNoneZeroErrorCode");
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java]: Why do you add it?

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/PrConfigView/config.jelly [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+         xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+
+
+    <f:entry title="${%Technologies Options}">
+        <f:hetero-list name="techOptions" items="${instance.techOptions}" descriptors="${descriptor.technologiesDescriptors}"
+                       hasHeader="true" oneEach="true"/>
+    </f:entry>
+
+    <f:entry title="App Name" field="appName">
+        <f:textbox value="${appName}" default="$${JOB_NAME}"/>
```
Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/PrConfigView/config.jelly]: Please show that to Jason and verify that the order and caption of the fields make sense.

Comment (nadavye) [reply src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/PrConfigView/config.jelly]: What about help file that describes each field and where does the user get it?

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.mockito.Mockito.mock;
+
+public class PrConfigTest extends ConfigTest {
+    private String latestCommit = "latestCommit";
+    private String pullRequestNumber = "1";
```
Comment (nadavye) [inline src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java]: Please add "final" before each string and write them in UPPER_CASE as they are constants.

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    private PrConfigCommandArguments createPrConfigArguments(DescribableList<TechnologyOptions,
+     TechnologyOptionsDescriptor> technologyOptions) {
+        PrConfigCommandArguments configArguments = new PrConfigCommandArguments(technologyOptions, latestCommit,
+                pullRequestNumber, repoUrl, targetBranch);
+
+        return configArguments;
+    }
+
+    private DescribableList<TechnologyOptions, TechnologyOptionsDescriptor> createTechnologyOptions() {
```
Comment (nadavye) [inline src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java]: Seems like no one is using this code. Am I'm missing something?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIHandler.java (hunk 1 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIHandler.java]: This is a very ugly code which coupled the upgrade logic and relevant jar, for config and prCoonfig modes we need the build scanner, all other uses test listener.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner.java (hunk 2 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CLIRunner.java]: Done

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java (hunk 3 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: File not found

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java (hunk 6 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java]: No its used in jelly files, tried before and didn't work

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java (hunk 7 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java]: Yes, token data resolved in the beginning of ```CLIRunner```

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java (hunk 8 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigCommandExecutor.java]: Its required by agent

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java (hunk 11 above)
Comment (RoniSegal) [reply src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java]: Done

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java (hunk 10 above)
Comment (RoniSegal) [reply src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/PrConfigTest.java]: Done

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java (hunk 5 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/PrConfigCommandArguments.java]: Done

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/PrConfigView/config.jelly (hunk 9 above)
Comment (RoniSegal) [reply src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode/PrConfigView/config.jelly]: I'm waiting for Marc he should provide the exec git commands

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java (hunk 3 above)
Comment (nadavye) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/CommandMode.java]: Sent you in slack.

**Pull request merged**